rfc/007-workload-secrets.md

```diff
@@ -0,0 +1,108 @@
+After being provisioned with the unsealing key it can resume normal operation.
+When running Vault as a Contrast workload, access to a persistent secret would allow automatic unsealing without operator intervention.
+
+[Hashicorp Vault]: https://www.hashicorp.com/products/vault
+[OpenBao]: https://openbao.org/
+[storage sealing]: https://developer.hashicorp.com/vault/docs/concepts/seal
+
+## Proposal
+
+### Secret source
+
+Workload secrets would need to be either persisted or deterministically derived.
+The only secure persistence option for the Coordinator is encryption with a key derived from the secret seed.
+Since the security and access guarantees of such a scheme are directly tied to the seed, deriving the workload secret directly from the seed is the easier option, and more reliable.
+
+The derived key should be large enough to allow secure derivation of common asymmetric key material.
+Recommended[^1] for P-512 would be 48 bytes, and a secret key for X448 also needs 48 bytes.
+Since the maximum entropy is limited by the Coordinator's secret seed size, derive a key of the same size.
+Increasing the entropy of the Coordinator seed should then be considered separately.
+
+[^1]: https://pkg.go.dev/filippo.io/keygen#ECDSA
+
+### Workload identity
+
+Access to workload secrets should be tied to a _workload identity_.
+This identity needs to be stable for at least all pods sharing a runtime policy, because that's the finest-grained identity Contrast can verify.
+However, runtime policies aren't stable enough to satisfy the outlined conditions.
+For example, policy rules can change between Contrast versions, or when workload parameters change.
+Therefore, the identity can't be the policy hash, and similar arguments hold for other Contrast-managed identity schemes.
+
+Instead of deriving the identity from workload characteristics, users are going to label workloads with an identity.
+To allow that, we're' going to change the `Manifest` schema.
+The policies field of the manifest is currently a plain map of policy hashes to SANs.
+
+```json
+{
+    "Policies": {
+        "99dd77cbd7fe2c4e1f29511014c14054a21a376f7d58a48d50e9e036f4522f6b": [
+            "web",
+            "*",
+            "203.0.113.34"
+        ]
+    }
+}
+```
+
+Changing this to an object with named fields allows adding new metadata, for example a `WorkloadSecretID`.
+
+```json
+{
+    "Policies": {
+        "99dd77cbd7fe2c4e1f29511014c14054a21a376f7d58a48d50e9e036f4522f6b": {
+            "SAN": ["web", "*", "203.0.113.34"],
+            "WorkloadSecretID": "openbao-prod",
+            // ...
+        }
+    }
+}
+```
+
+### Implementation
+
+After successful workload verification, the Coordinator derives a key with HKDF, setting the info argument to `workload-key:$WorkloadSecretID`.
+This key is returned as a new field `WorkloadKey` in the `NewMeshCertResponse`.
+The initializer writes the key to a file in the shared tmpfs, from where it can be used by the workload.
```

Comment:
This part is similar to MarbleRun. In addition, MarbleRun allows to inject the secret at different places and in different formats. With this, you can often avoid adapting your workload. If we would want something similar for Contrast, I guess it should be build upon the primitive you introduce with this RFC, right? For example, one could create an init container that transforms the secret such that the actual workload wouldn't need to be modified. Do you think this works or is there anything we'd need to consider for this RFC to enable this use case?

Reply:
I guess it's a good sign that the proposals converge, even though I did not peek at the MarbleRun implementation :) 

The scheme proposed here is intentionally designed to support derivation of more keys, and these use cases should be enabled by it. 

The initializer (and other init containers) is somewhat limited in terms of what it can inject. Shared files are definitely possible, but env vars are not. I think the annotation pattern we use for the service mesh would be a good starting point for this.